pypicloud/access/_ldap.py

```diff
@@ -0,0 +1,292 @@
+
+    @staticmethod
+    def all_usernames():
+        """
+        Returns a list of all user names
+        """
+        if not hasattr(LDAP, "_all_users"):
+            LDAP._initialize_cache()
+        return list(LDAP._all_users.keys())
+
+    @staticmethod
+    def user_dn(username):
+        """
+        Returns the dn for the username
+        """
+        if not hasattr(LDAP, "_all_users"):
+            LDAP._initialize_cache()
+        return LDAP._all_users.get(username)
+
+    @staticmethod
+    @reconnect
+    def _add_admins_from_dn(admin_dn):
+        """
+        Given a DN fragement, add users to _admins and _admin_usernames
+        """
+        res = LDAP._server.search_s(admin_dn, ldap.SCOPE_SUBTREE)
+        try:
+            admins = res[0][1][LDAP._admin_field]
+        except (IndexError, KeyError) as err:
+            logging.warn("Error retrieving admins from %s: %r", admin_dn, err)
+            return
+
+        for admin in admins:
+            try:
+                LDAP._admin_usernames.append(
+                    LDAP._server.search_s(
+                        admin,
+                        ldap.SCOPE_SUBTREE,
+                    )[0][1][LDAP._id_field][0]
+                )
+            except (IndexError, KeyError) as error:
+                logging.warn("Error looking up admin %s: %r", admin, error)
+            else:
+                LDAP._admins.append(admin)
+
+    @staticmethod
+    def admins():
+        """
+        Returns a list of all the admin DNs
+        """
+        if not hasattr(LDAP, "_admins"):
+            LDAP._admins = []
+            LDAP._admin_usernames = []
+            for admin_dn in LDAP._admin_dns:
+                LDAP._add_admins_from_dn(admin_dn)
+
+        return LDAP._admins
+
+    @staticmethod
+    def admin_usernames():
+        """
+        Returns a list of the admin usernames
+        """
+        if not hasattr(LDAP, "_admin_usernames"):
+            LDAP.admins()
+        return LDAP._admin_usernames
+
+    @staticmethod
+    @reconnect
+    def bind_user(user_dn, password):
+        """
+        Attempts to bind as the user, then rebinds as service user again
+        """
+        try:
+            LDAP._server.simple_bind_s(user_dn, password)
+        except ldap.INVALID_CREDENTIALS:
+            return False
+        else:
+            return True
+        finally:
+            LDAP._connect()
+
+
+class LDAPAccessBackend(IAccessBackend):
+    """
+    This backend allows you to authenticate against a remote LDAP server.
+    """
+
+    def __init__(self, request=None, group_map=None, **kwargs):
+        super(LDAPAccessBackend, self).__init__(request, **kwargs)
+        self.group_map = group_map
+
+    @classmethod
+    def configure(cls, settings):
+        kwargs = super(LDAPAccessBackend, cls).configure(settings)
+        LDAP.configure(settings)
+        kwargs["group_map"] = {
+            "admin": ("read", "write"),
+            "authenticated": ("read",),
```

Comment:
they're usually not needed yeah. but if someone configures default_write to some other group, without those lines, admin would lose write access. same with authenticated and the inherent default read.